### PR TITLE
Continue bootstrapmethods parsing after unsupported entry

### PR DESCRIPTION
**Pull request by mgudemann**
Before it was possible to keep some `BootstrapMethods` entries unparsed, after a
currently unsupported was found. This lead to the situation that JBMC ended up
with a wrong file offset which caused problems in the case of class file
attributes being present after the `BootstrapMethods` attribute.

This was not observed using the Oracle or OpenJDK compilers, but currently only
on the Eclipse compiler for Java.

### CODE REVIEW

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_parser.cpp]: Why has this changed from `i` to `j`

Comment (thk123) [reply src/java_bytecode/java_bytecode_parser.cpp]: Because it was colliding with another loop, perhaps a good argument for a better name like `boostrap_method_index`?

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_parser.cpp]: It is a bit hard to follow these nested if/else statements, consider changing to:
```c++
if(! /* validity check 1 */) 
{
  report_error1();
  continue;
}
if(! /* validity check 2 */) 
{
  report_error2();
  continue;
}
// do the real work!
```

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc
Comment (thk123) [inline regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: Is this compiled with eclipse compiler? If so probably worth making a note of that here to stop people inadvertently updating it.

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (majakusber) [reply src/java_bytecode/java_bytecode_parser.cpp]: +1 Some commentary on the individual cases would make it easier to follow too.

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc
Comment (majakusber) [inline regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: Why this method is tested instead of `Smr`? Wouldn't `Smr` do the same job, with lambdas being ignored but parsed?

**Review by majakusber — CHANGES_REQUESTED** [review]
The if-else statements are nested now which makes them hard to read. It would be useful for future work on lambdas to make it more readable.

**Comment by thk123** [comment]
I've done as suggested and replaced the nested ifs with `if(!valid) continue;` style code as well as pulling out the error producing code into a separate function to reduce duplication. I've also tried to document each of the errors, so @svorenova could you see if they make sense? Will consider the tests next.

**Comment by majakusber** [comment]
@thk123 Yes, looks very good, thank you! Will wait for the test to approve.

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc
Comment (thk123) [reply regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: It appears to be - will update docs

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc
Comment (thk123) [reply regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: I agree and the test passes setting to the the `Smr` so will change to that but @mgudemann when you're back if you could comment on whether this was the right thing to do.

**Comment by thk123** [comment]
@svorenova Adjusted the tests

File: regression/cbmc-java/lambda2/StaticMethodRef.java
Comment (majakusber) [inline regression/cbmc-java/lambda2/StaticMethodRef.java]: This can be removed now it seems.

**Review by majakusber — APPROVED** [review]
Looks good, just one line in the test to remove but non-blocking.

**Review by thk123 — APPROVED** [review]
I've applied my comments, so I'm happy, probably good to get a second review from a different "senior" e.g. @allredj ?

**Comment by allredj** [comment]
@thk123 OK I'll take a look.

File: regression/cbmc-java/lambda2/StaticMethodRef.java
Comment (allredj) [inline regression/cbmc-java/lambda2/StaticMethodRef.java]: Nit: Java format is not uniform across the file.

File: regression/cbmc-java/lambda2/StaticMethodRef.java
Comment (allredj) [inline regression/cbmc-java/lambda2/StaticMethodRef.java]: Nit: extra spaces in the generic signature.

File: regression/cbmc-java/lambda2/StaticMethodRef.java
Comment (allredj) [inline regression/cbmc-java/lambda2/StaticMethodRef.java]: Bad indent

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (allredj) [inline src/java_bytecode/java_bytecode_parser.cpp:1906]: Question for my understanding: Why do we want to keep these unknown handles instead of just skipping the?

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (allredj) [inline src/java_bytecode/java_bytecode_parser.cpp]: `boostrap` -> `bootstrap`

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (allredj) [inline src/java_bytecode/java_bytecode_parser.cpp]: Nit: Don't we have an 80-char limit?

Comment (allredj) [reply src/java_bytecode/java_bytecode_parser.cpp]: The way we access this map looks a bit unnatural. Why does name of `parsed_class` have to be an element of the map entry, that itself belongs to the same class. Obviously the name will always be identical for a given `parsed_class` or am I missing something? Could we have a wrapper method in `java_bytecode_parse_treet::classt` that cares about adding an element to the map instead of accessing the map directly?

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (allredj) [inline src/java_bytecode/java_bytecode_parser.cpp]: Nit: I find conditions of the form:
```
interface_type_argument.tag != CONSTANT_MethodType ||
method_handle_argument.tag != CONSTANT_MethodHandle ||
method_type_argument.tag != CONSTANT_MethodType
```
easier to read

**Review by allredj — APPROVED** [review]
Looks fine as far as I can tell. I have a couple of questions.

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc
Comment (allredj) [inline regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: That's a bit confusing for whoever will see this test fails once lambdas start working. Could we add a comment saying what specifically we are testing here?

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [reply src/java_bytecode/java_bytecode_parser.cpp:1906]: I think because the `invokedynamic` refers to an index in the bootstrap table which would be invalid if we didn't populate all the entries .

File: src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [reply src/java_bytecode/java_bytecode_parser.cpp]: Yes I agree  tagging @mgudemann to specifically review what I've done here but I agree this info is redundant and a method will make updating this in the future simpler - will do.

**Comment by thk123** [comment]
@allredj Addressed all comments, though have spun out as a separate ticket to make the bootstrap methods map itself private as a bit more involved

**Comment by allredj** [comment]
@thk123 thanks!

**Pull request merged**
